File: screen_observer_client.py
"""
screen_observer_client.py — Optional HTTP client for OS Screen Observer.

Used as a perception overlay in backends/base.py when screen_observer.enabled=true
in config.json.  All methods return None on any network/server failure so callers
can fall through to their native perception path.  A 30-second cooldown prevents
hammering a down server.
"""

import logging
import time

logger = logging.getLogger(__name__)

_COOLDOWN = 30.0

# Capability keys from /api/capabilities that AutoGUI uses.
# Defaults apply when OSO is reachable but the key is absent (older OSO version).
_CAP_DEFAULTS: dict = {
    "accessibility_tree": True,
    "ocr": True,
    "vlm": False,
    "screenshot": True,
    "monitors": True,
}


def _oso_walk_for_element(
    node: dict,
    name_query: str,
    control_type: str | None,
) -> list[dict]:
    """Recursively flatten OSO element tree into matching elements."""
    results = []
    name_lower = name_query.lower()
    node_name = (node.get("name") or "").lower()
    node_role = node.get("role", "")

    name_match = name_lower in node_name
    type_match = control_type is None or control_type.lower() in node_role.lower()

    if name_match and type_match and node.get("bounds"):
        b = node["bounds"]
        results.append({
            "name": node.get("name", ""),
            "control_type": node_role,
            "rect": {
                "x": b.get("x", 0),
                "y": b.get("y", 0),
                "width": b.get("width", 0),
                "height": b.get("height", 0),
            },
            "method": "screen_observer",
        })

    for child in node.get("children", []):
        results.extend(_oso_walk_for_element(child, name_query, control_type))

    return results


class ScreenObserverClient:
    """
    Thin async HTTP client for the OS Screen Observer REST API
    (http://127.0.0.1:5001 by default).

    All public methods are async and return None when the server is
    unreachable or returns an error, so callers can always fall back
    to their native path without extra error handling.
    """

    def __init__(self, cfg: dict):
        self._base = cfg.get("base_url", "http://127.0.0.1:5001").rstrip("/")
        self._timeout = float(cfg.get("timeout_seconds", 2.0))
        self._enabled = bool(cfg.get("enabled", False))
        self._disabled_until: float = 0.0
        # Cached result of /api/capabilities; populated on first is_available() call.
        self._caps: dict = dict(_CAP_DEFAULTS)
        self._caps_fetched: bool = False

    @property
    def enabled(self) -> bool:
        return self._enabled

    @property
    def oso_capabilities(self) -> dict:
        """Return the last-fetched /api/capabilities 'supports' dict.

        Available after the first successful is_available() or get_capabilities()
        call.  Falls back to conservative defaults until then.
        """
        return dict(self._caps)

    def _cooled(self) -> bool:
        return time.monotonic() >= self._disabled_until

    def _back_off(self) -> None:
        self._disabled_until = time.monotonic() + _COOLDOWN

    async def _get(self, path: str, params: dict | None = None) -> dict | None:
        if not self._enabled or not self._cooled():
            return None
        try:
            import aiohttp
            async with aiohttp.ClientSession() as s:
                async with s.get(
                    f"{self._base}{path}",
                    params=params or {},
                    timeout=aiohttp.ClientTimeout(total=self._timeout),
                ) as r:
                    if r.status == 200:
                        return await r.json()
                    logger.warning("[OSO] GET %s -> HTTP %d; falling back to native method", path, r.status)
                    return None
        except Exception as e:
            logger.warning("[OSO] GET %s failed: %s; falling back to native method", path, e)
            self._back_off()
            return None

    async def is_available(self) -> bool:
        """Probe /api/healthz; returns True if the server is reachable.

        Also fetches /api/capabilities on the first successful probe so
        subsequent calls to oso_capabilities reflect what this OSO instance
        actually supports.
        """
        result = await self._get("/api/healthz")
        if result is not None and not self._caps_fetched:
            await self.get_capabilities()
        return result is not None

    async def get_capabilities(self) -> dict | None:
        """Fetch /api/capabilities and cache the 'supports' dict.

        Returns the full capabilities response, or None if unreachable.
        The 'supports' sub-dict is always accessible via oso_capabilities
        even after a failure (returns cached or default values).
        """
        result = await self._get("/api/capabilities")
        if result is not None:
            supports = result.get("supports") or {}
            merged = dict(_CAP_DEFAULTS)
            merged.update(supports)
            self._caps = merged
            self._caps_fetched = True
            logger.info(
                "[OSO] capabilities: version=%s supports=%s",
                result.get("version", "?"),
                self._caps,
            )
        return result

    async def get_windows(self) -> dict | None:
        return await self._get("/api/windows")

    async def get_description(self, window_index: int | None = None) -> dict | None:
        p: dict = {}
        if window_index is not None:
            p["window_index"] = window_index
        return await self._get("/api/description", p)

    async def get_sketch(self, window_index: int | None = None) -> dict | None:
        p: dict = {}
        if window_index is not None:
            p["window_index"] = window_index
        return await self._get("/api/sketch", p)

    async def get_structure(self, window_index: int | None = None) -> dict | None:
        p: dict = {}
        if window_index is not None:
            p["window_index"] = window_index
        return await self._get("/api/structure", p)

    async def get_screenshot(self, window_index: int | None = None) -> dict | None:
        """Fetch a screenshot from OSO (/api/screenshot).

        Returns {data: base64_png, format: 'png', encoding: 'base64', window: title}
        or None on failure.  Useful for per-window captures that OSO performs
        natively via its accessibility adapter.
        """
        if not self._caps.get("screenshot", True):
            return None
        p: dict = {}
        if window_index is not None:
            p["window_index"] = window_index
        return await self._get("/api/screenshot", p)

    async def get_monitors(self) -> dict | None:
        """Fetch monitor geometry from OSO (/api/monitors).

        Returns {ok, monitors: [{bounds, scale_factor, ...}]} or None on failure.
        Useful for multi-monitor coordinate translation and DPI-aware positioning.
        """
        if not self._caps.get("monitors", True):
            return None
        return await self._get("/api/monitors")

    async def get_visible_areas(self, window_index: int) -> dict | None:
        """Fetch the unoccluded regions of a window (/api/visible_areas).

        Returns {window, visible_regions: [{x, y, width, height}]} or None.
        Useful for verifying a click target is reachable without hitting an
        overlapping window.
        """
        return await self._get("/api/visible_areas", {"window_index": window_index})

    async def find_element_in_tree(
        self,
        name: str,
        control_type: str | None = None,
        window_index: int | None = None,
        index: int = 0,
    ) -> dict | None:
        """Get the structure tree and walk it for a matching element."""
        tree_result = await self.get_structure(window_index=window_index)
        if tree_result is None:
            return None
        matches = _oso_walk_for_element(
            tree_result.get("tree") or {},
            name_query=name,
            control_type=control_type,
        )
        if not matches:
            return None
        idx = max(0, min(int(index or 0), len(matches) - 1))
        return matches[idx]

File: pi-extension/src/screen_observer_client.ts
/**
 * screen_observer_client.ts — Optional HTTP/MCP client for OS Screen Observer.
 *
 * Port of mainline `screen_observer_client.py`.  Used by tools.ts as a
 * perception overlay when an OSO server is reachable (auto-probed in
 * index.ts).  All methods return null on any failure so callers fall
 * through to the native backend path with no extra error handling.
 *
 * A 30-second cooldown prevents hammering a down server.  An MCP-backed
 * variant is supplied by index.ts when an OSO MCP server is registered
 * with the Pi runtime; the surface is identical so tools.ts doesn't care
 * which transport is in use.
 */
import type { Logger } from "./logger.js";

const COOLDOWN_MS = 30_000;

const noopLogger: Logger = { async log() { /* no-op */ } };

/** Capability keys from /api/capabilities 'supports' dict.  Defaults apply
 *  when OSO is reachable but the key is absent (older OSO versions).
 *  Mirror of `_CAP_DEFAULTS` in screen_observer_client.py. */
export const OSO_CAPABILITY_DEFAULTS: Record<string, boolean> = {
  accessibility_tree: true,
  ocr: false,
  vlm: false,
  uia_invoke: false,
  occlusion_detection: true,
  drag: true,
  screenshot: true,
  monitors: true,
  bring_to_foreground: true,
  element_targeting: true,
  observe_with_diff: true,
};

export interface OsoConfig {
  base_url?: string;
  timeout_seconds?: number;
  enabled?: boolean;
}

export interface OsoElement {
  element_id?: string;
  name: string;
  control_type: string;
  rect: { x: number; y: number; width: number; height: number };
  method: string;
}

interface TreeNode {
  name?: string;
  role?: string;
  bounds?: { x?: number; y?: number; width?: number; height?: number };
  children?: TreeNode[];
}

/** Recursively flatten OSO element tree into matching elements. */
function walkForElement(node: TreeNode | undefined, nameQuery: string, controlType: string | null): OsoElement[] {
  if (!node) return [];
  const results: OsoElement[] = [];
  const nameLower = nameQuery.toLowerCase();
  const nodeName = (node.name ?? "").toLowerCase();
  const nodeRole = node.role ?? "";
  const nameMatch = nameLower.length === 0 || nodeName.includes(nameLower);
  const typeMatch = !controlType || nodeRole.toLowerCase().includes(controlType.toLowerCase());
  if (nameMatch && typeMatch && node.bounds) {
    const b = node.bounds;
    results.push({
      name: node.name ?? "",
      control_type: nodeRole,
      rect: {
        x: b.x ?? 0,
        y: b.y ?? 0,
        width: b.width ?? 0,
        height: b.height ?? 0,
      },
      method: "screen_observer",
    });
  }
  for (const child of node.children ?? []) {
    results.push(...walkForElement(child, nameQuery, controlType));
  }
  return results;
}

/**
 * Transport seam: the REST client uses `fetch`; the MCP-backed variant in
 * index.ts implements the same interface by routing each method to an
 * MCP tool call.  Keeping the type abstract lets tools.ts depend on the
 * surface only, not the transport.
 */
export interface ScreenObserverClient {
  readonly enabled: boolean;
  readonly osoCapabilities: Record<string, boolean>;
  isAvailable(): Promise<boolean>;
  getCapabilities(): Promise<Record<string, unknown> | null>;
  getWindows(): Promise<Record<string, unknown> | null>;
  getScreenshot(windowIndex?: number): Promise<Record<string, unknown> | null>;
  getMonitors(): Promise<Record<string, unknown> | null>;
  getStructure(windowIndex?: number): Promise<{ tree?: TreeNode } | null>;
  findElement(query: { name?: string; controlType?: string; windowTitle?: string; windowIndex?: number; index?: number }): Promise<OsoElement | null>;
  findElementInTree(query: { name: string; controlType?: string; windowIndex?: number; index?: number }): Promise<OsoElement | null>;
  bringToForeground(query: { windowTitle?: string; windowIndex?: number; windowUid?: string }): Promise<Record<string, unknown> | null>;
  observe(query?: { windowIndex?: number; windowTitle?: string }): Promise<Record<string, unknown> | null>;
  elementClick(elementId: string): Promise<Record<string, unknown> | null>;
  elementFocus(elementId: string): Promise<Record<string, unknown> | null>;
  elementInvoke(elementId: string): Promise<Record<string, unknown> | null>;
  elementSetValue(elementId: string, value: string): Promise<Record<string, unknown> | null>;
}

export class RestScreenObserverClient implements ScreenObserverClient {
  private base: string;
  private timeoutMs: number;
  private _enabled: boolean;
  private disabledUntil = 0;
  private caps: Record<string, boolean> = { ...OSO_CAPABILITY_DEFAULTS };
  private capsFetched = false;

  private logger: Logger;

  constructor(cfg: OsoConfig, logger: Logger = noopLogger) {
    this.base = (cfg.base_url ?? "http://127.0.0.1:5001").replace(/\/+$/, "");
    this.timeoutMs = Math.round((cfg.timeout_seconds ?? 2.0) * 1000);
    this._enabled = !!cfg.enabled;
    this.logger = logger;
  }

  get enabled(): boolean { return this._enabled; }
  get osoCapabilities(): Record<string, boolean> { return { ...this.caps }; }

  /** Force-enable after a successful out-of-band probe. */
  enable(): void { this._enabled = true; }

  private cooled(): boolean { return Date.now() >= this.disabledUntil; }
  private backOff(): void { this.disabledUntil = Date.now() + COOLDOWN_MS; }

  private async request(method: "GET" | "POST", path: string, params?: Record<string, unknown>, body?: unknown): Promise<Record<string, unknown> | null> {
    if (!this._enabled || !this.cooled()) return null;
    let url = `${this.base}${path}`;
    if (method === "GET" && params && Object.keys(params).length > 0) {
      const usp = new URLSearchParams();
      for (const [k, v] of Object.entries(params)) {
        if (v === undefined || v === null) continue;
        usp.append(k, String(v));
      }
      const q = usp.toString();
      if (q) url += `?${q}`;
    }
    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), this.timeoutMs);
    try {
      const res = await fetch(url, {
        method,
        signal: ctrl.signal,
        headers: body !== undefined ? { "content-type": "application/json" } : undefined,
        body: body !== undefined ? JSON.stringify(body) : undefined,
      });
      if (!res.ok) {
        await this.logger.log("oso.http_error", { path, status: res.status });
        return null;
      }
      return await res.json() as Record<string, unknown>;
    } catch (e) {
      await this.logger.log("oso.request_failed", { path, error: (e as Error).message });
      this.backOff();
      return null;
    } finally {
      clearTimeout(timer);
    }
  }

  async isAvailable(): Promise<boolean> {
    const r = await this.request("GET", "/api/healthz");
    if (r !== null && !this.capsFetched) {
      await this.getCapabilities();
    }
    return r !== null;
  }

  async getCapabilities(): Promise<Record<string, unknown> | null> {
    const r = await this.request("GET", "/api/capabilities");
    if (r !== null) {
      const supports = (r.supports as Record<string, boolean>) ?? {};
      this.caps = { ...OSO_CAPABILITY_DEFAULTS, ...supports };
      this.capsFetched = true;
      await this.logger.log("oso.capabilities", { version: r.version ?? "?", supports: this.caps });
    }
    return r;
  }

  getWindows(): Promise<Record<string, unknown> | null> { return this.request("GET", "/api/windows"); }

  getScreenshot(windowIndex?: number): Promise<Record<string, unknown> | null> {
    return this.request("GET", "/api/screenshot", windowIndex !== undefined ? { window_index: windowIndex } : undefined);
  }

  getMonitors(): Promise<Record<string, unknown> | null> { return this.request("GET", "/api/monitors"); }

  getStructure(windowIndex?: number): Promise<{ tree?: TreeNode } | null> {
    return this.request("GET", "/api/structure", windowIndex !== undefined ? { window_index: windowIndex } : undefined) as Promise<{ tree?: TreeNode } | null>;
  }

  async findElementInTree(query: { name: string; controlType?: string; windowIndex?: number; index?: number }): Promise<OsoElement | null> {
    const tree = await this.getStructure(query.windowIndex);
    if (!tree) return null;
    const matches = walkForElement(tree.tree, query.name, query.controlType ?? null);
    if (matches.length === 0) return null;
    const idx = Math.max(0, Math.min(query.index ?? 0, matches.length - 1));
    return matches[idx];
  }

  async findElement(query: { name?: string; controlType?: string; windowTitle?: string; windowIndex?: number; index?: number }): Promise<OsoElement | null> {
    if (this.caps.accessibility_tree === false) return null;
    if (!query.name && !query.controlType) return null;
    const role = query.controlType ?? "*";
    const selector = query.name ? `//${role}[name*="${query.name}"]` : `//${role}`;
    const params: Record<string, unknown> = { selector };
    if (query.windowIndex !== undefined) params.window_index = query.windowIndex;
    else if (query.windowTitle) params.window_title = query.windowTitle;
    const r = await this.request("GET", "/api/find_element", params);
    if (!r || !r.ok) return null;
    const allMatches = (r.all_matches as Array<Record<string, unknown>> | undefined) ?? [];
    if (allMatches.length > 0 && (query.index ?? 0) > 0) {
      const idx = Math.max(0, Math.min(query.index!, allMatches.length - 1));
      const m = allMatches[idx];
      return {
        element_id: m.element_id as string | undefined,
        name: (m.name as string) ?? "",
        control_type: (m.role as string) ?? "",
        rect: (m.bounds as OsoElement["rect"]) ?? { x: 0, y: 0, width: 0, height: 0 },
        method: "screen_observer_find_element",
      };
    }
    const bounds = (r.bounds as OsoElement["rect"]) ?? { x: 0, y: 0, width: 0, height: 0 };
    const first = (allMatches[0] ?? {}) as Record<string, unknown>;
    return {
      element_id: r.element_id as string | undefined,
      name: (first.name as string) ?? "",
      control_type: (first.role as string) ?? "",
      rect: bounds,
      method: "screen_observer_find_element",
    };
  }

  bringToForeground(query: { windowTitle?: string; windowIndex?: number; windowUid?: string }): Promise<Record<string, unknown> | null> {
    const params: Record<string, unknown> = {};
    if (query.windowUid) params.window_uid = query.windowUid;
    else if (query.windowIndex !== undefined) params.window_index = query.windowIndex;
    else if (query.windowTitle) params.window_title = query.windowTitle;
    else return Promise.resolve(null);
    return this.request("GET", "/api/bring_to_foreground", params);
  }

  observe(query?: { windowIndex?: number; windowTitle?: string }): Promise<Record<string, unknown> | null> {
    const params: Record<string, unknown> = {};
    if (query?.windowIndex !== undefined) params.window_index = query.windowIndex;
    else if (query?.windowTitle) params.window_title = query.windowTitle;
    return this.request("GET", "/api/observe", params);
  }

  elementClick(id: string): Promise<Record<string, unknown> | null> { return this.request("POST", "/api/element/click", undefined, { element_id: id }); }
  elementFocus(id: string): Promise<Record<string, unknown> | null> { return this.request("POST", "/api/element/focus", undefined, { element_id: id }); }
  elementInvoke(id: string): Promise<Record<string, unknown> | null> { return this.request("POST", "/api/element/invoke", undefined, { element_id: id }); }
  elementSetValue(id: string, value: string): Promise<Record<string, unknown> | null> { return this.request("POST", "/api/element/set_value", undefined, { element_id: id, value }); }
}

/**
 * Probe order: caller-provided `mcpClient` first (MCP runtime), then the
 * REST endpoint at `cfg.base_url`.  Returns the first reachable client or
 * null if both fail.  Either way, the cooldown ensures repeat probes
 * don't stall startup.
 */
export async function probeScreenObserver(
  cfg: OsoConfig,
  logger: Logger = noopLogger,
  mcpProbe?: () => Promise<ScreenObserverClient | null>,
): Promise<ScreenObserverClient | null> {
  if (mcpProbe) {
    try {
      const mcp = await mcpProbe();
      if (mcp) return mcp;
    } catch (e) {
      await logger.log("oso.mcp_probe_failed", { error: (e as Error).message });
    }
  }
  const rest = new RestScreenObserverClient({ ...cfg, enabled: true }, logger);
  if (await rest.isAvailable()) return rest;
  return null;
}
